server/list.go

```diff
@@ -15,65 +17,89 @@ const (
 	GetItem(itemID string) (*Item, error)
 	RemoveItem(itemID string) error
 
-	GetItemOrder(userID string, itemID string, listID string) (*OrderElement, int, error)
-	GetItemListAndOrder(userID string, itemID string) (string, *OrderElement, int)
+	// GetItemOrder gets the item Order Element and position of the item itemID on user userID's list listID
+	GetItemOrder(userID, itemID, listID string) (*OrderElement, int, error)
+	// GetItemListAndOrder gets the item list, Order Element and position for user userID
+	GetItemListAndOrder(userID, itemID string) (string, *OrderElement, int)
+
+	// Order Element related functions
 
-	Add(userID string, itemID string, listID string, foreignUserID string, foreignItemID string) error
-	Remove(userID string, itemID string, listID string) error
-	Pop(userID string, listID string) (*OrderElement, error)
+	// Add creates a new OrderElement with the itemID, foreignUSerID and foreignItemID, and stores it
+	// on the listID for userID.
+	Add(userID, itemID, listID, foreignUserID, foreignItemID string) error
+	// Remove removes the OrderElement for itemID in listID for userID
+	Remove(userID, itemID, listID string) error
+	// Pop removes the first OrderElement in listID for userID and return it
+	Pop(userID, listID string) (*OrderElement, error)
 
-	GetList(userID string, listID string) ([]*OrderElement, error)
+	// GetList returns the list of OrderElement in listID for userID
+	GetList(userID, listID string) ([]*OrderElement, error)
 }
 
 type listManager struct {
-	store       ListStore
-	getUserName func(string) string
+	store ListStore
+	api   plugin.API
 }
 
 // NewListManager creates a new listManager
-func NewListManager(store ListStore, getUserName func(string) string) *listManager {
+func NewListManager(store ListStore, api plugin.API) *listManager {
 	return &listManager{
-		store:       store,
-		getUserName: getUserName,
+		store: store,
+		api:   api,
 	}
 }
 
-func (l *listManager) Add(userID string, message string) error {
+func (l *listManager) Add(userID, message string) error {
 	item := newItem(message)
 
-	l.store.AddItem(item)
+	if err := l.store.AddItem(item); err != nil {
+		return err
+	}
 
 	if err := l.store.Add(userID, item.ID, MyListKey, "", ""); err != nil {
+		l.store.RemoveItem(item.ID)
 		return err
 	}
 
 	return nil
 }
 
-func (l *listManager) Send(senderID string, receiverID string, message string) (string, error) {
+func (l *listManager) Send(senderID, receiverID, message string) (string, error) {
 	senderItem := newItem(message)
-	l.store.AddItem(senderItem)
+	if err := l.store.AddItem(senderItem); err != nil {
+		return "", err
+	}
+
 	receiverItem := newItem(message)
-	l.store.AddItem(receiverItem)
+	if err := l.store.AddItem(receiverItem); err != nil {
+		l.store.RemoveItem(senderItem.ID)
+		return "", err
+	}
 
 	appErr := l.store.Add(senderID, senderItem.ID, OutListKey, receiverID, receiverItem.ID)
 	if appErr != nil {
+		l.store.RemoveItem(senderItem.ID)
```

Comment:
Here I am not dealing with the errors on purpose. There is nothing to handle if something fails, since this is some kind of "rollback" of the changes. The only think I can do is keep trying to remove the item until it gets erased, but since it may fail for other things, it may not be the best approach.

Another possibility would be to include these IDs to a "Garbage collector" to retry the remove later, but that feels a bit of a overkill.

Reply:
I just made some changes regarding this, but I am not sure how good they are. The thing is that many errors are dealt with "silently", as I explained in my previous comment. But it might be interesting to have them logged to see if anything is going wrong. I struggled whether to put them as Error or as Debug, but I ended up setting them as errors.

Also, I do not feel specially comfortable with how the code looks now after the changes regarding this, but right now I cannot think of a better way to log these errors. What do you think?

Reply:
I think this is fine -- it also helps with documenting what all the calls are supposed to be doing. I found it helpful for that reason alone. :) 

Also, the comments you added to the functions are very helpful as well. Thanks!